### ISSUE TITLE
Cannot copy Date type via CamelCasedPropertiesDeep

### ISSUE THREAD

**Issue by David245M** [sindresorhus/type-fest#326]
I'm currently using `CamelCasedPropertiesDeep` type and I found that it's not copy Date type, instead it creates copy of it. I've got this error
```
Argument of type '{ toString: () => string; toDateString
string; toLocaleString: { (): string; (locales?: string | string[] | undefined, options?: DateTimeFormatOptions | undefineON: (key?: any) => string; }' is not assignable to parameter of type 'number | Date'.
```
So I tried it with another standard build-in object types and it works with none of them.
```
interface ExampleType {
    bool: Boolean
    sym: Symbol
    error: Error
    num: Number
    bigInt: BigInt
    math: Math
    date: Date
    str: String
    regExp: RegExp
    array: Int8Array
    map: Map<any,any>
    weakMap: WeakMap<any, any>
    weakSet: WeakSet<any>
    json: JSON
    promise: Promise<any>
    generator: Generator
    dateTimeFormat: Intl.DateTimeFormat
}

type CamelizedExample = CamelCasedPropertiesDeep<ExampleType>
```
[Here](https://www.typescriptlang.org/play?target=8#code/C4TwDgpgBAymA2BLYAeAsAKAJAyhAHsBAHYAmAzlOcAE6LEDmANJlgCIRIC2yENehEhSq16zTAD4oAXlgCiZSgAMAJAG96AMz5QAEhACGpAL7qO3XjVMbi2-gBUDieMaWsA-FADa+o0ygAdEFwSKiOzv7miDxENBIAuqwAXHIECsJRMXwe3onYKV4w8QDcmJgEYAD2NMBQoJBQAOrVpDAQYAY0BsDVlLIA5AC0-VAAPlD9APoj4yP9pRhlGPXQAJLExHwAwgZcnDvkEDCijACCNF0gKAAKncCUaUKUNIaklcTwIFAGxCBe8f5ri8AG6ISoAV3ItxqUmkrGh93kT28qhsdigADFEDRqAjXP4ggEtDoAEoQLhOYhiBHkPJYLCeLE44AIpGKKDgsgQTT0CCkVj0zz9foCrApJm4u5s4TC0WCqDrTY0HZ7eAHI4nBjnS4oMkU+jUu7kQEgsGQhESOViqCooEQUEQqFSx7s4VQRnYyU1KApHZgZAGJAALwgKAlLLuEmsiu2u32BkOxzoZwuBiuesphpqxsxnojMNcApSwoWmBWUBV8cTmu1aZuRulz1e70+ImTDH+sPhDZdwi8qOJ-HDeKUBKCg6gGYNjBpdIZUAAqsQAMYGf3AQOIEMoVHDu7Rjax1XqpNiWvp8mZmdG-x7gtKS3Ya2bYF8UuLZbgaCVtUJ0MAaVhKB-0bNsxHdCs41-at23PFAQmQFAQN7SgFzASAaFXQ4kKkTwABlKgAdz4LCAKkFJ-38ZoaFadpOm6XoJHI4D33LH91VIIFKgw4BEAgcgOHaFAADVA3BCAgNE+BxNAjFOWXXj3hyKTxOSKAVOgFCoDgicF0fec4PYv9OJobi+F4-jBLAFA9P0lINNAtpUF0-TPCclAjMOEyzJqPiBIgITbJ9KA1FFLwQPoKAAGsIBASpNHUsToATSDjz-HD4l9KCOK4ni-KskSkvC+J9OMd96FiTQDGXaAAFF8F2BAIHsL8QswKAOqgAAjSpKngFIACFevgQxiHazryBALgUhgKaevgcaOr4UyaBSWqLmqRaoGIcFpqgAA5XauuyDBOu6xAGHWYBBouq6topYAAAsUgAWW6R6ttIboIBSNhvq26hVtgTUtpeBh6rAFIyXB-AwC2zpLhSK6AA5z3utdXrXFAfhAJgccfM7iIMKK3shppDBJrGcf8fGtqJqKnJSRoKfc2nTs6gArch3hSAApGAAHl9q2sBTJ4Q4Ui48XQzZs6GBIPgGKBgBxBWuh6GhPu++xoggDFqgepHiGAeAAj+ogdb2fWaAezBjCWNioK3Pl6sakaZFSqs+Vy8z8oC6zXa4JqWsgR8gA) you can try to camelize keys of an object I've present above and notice this issue.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/326">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/326/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/sindresorhus/type-fest/issues/326/pledge.svg">
</picture>
</a>

**Comment by voxpelli** [comment]
I think easiest solution here would be to add a new `exclude` parameter like `CamelCasedPropertiesDeep<ExampleType, String | Date>`

Some of these should be getting built in support though, like `Promise`, `Map` and maybe `Date`.

**Comment by David245M** [comment]
@voxpelli I'm thinking about add check like
```
 : Value extends Map<infer Key, infer Val>
 ? Map<CamelCasedPropertiesDeep<Key>, Val>
 : Value extends StandartBuildInTypes
 ? Value
 : {
    ...
 }
 ```
 This will copy type instead of iterating it and creating the new ones. `StandartBuildInTypes` is just  the union of standart types excluding `Object` and iterative types. 
 What do you think? I've searching for the way to distinguish types like `Date` and `RegExp` from `Object` for the whole day, but I haven't found any good way except I've describe above.

**Comment by voxpelli** [comment]
A problem with that way is that TypeScript defines objects are that its pure duck typing: As long as it looks like a duck, quacks like a duck, it is a duck.

See eg [this playground](https://www.typescriptlang.org/play?target=8#code/C4TwDgpgBAymA2BLYAeAsAKAJAyhAHsBAHYAmAzlOcAE6LEDmANJlgCIRIC2yENehEhSq16zTAD4oAXlgCiZSgAMAJAG96AMz5QAEhACGpAL7qO3XjVMbi2-gBUDieMaWsA-FADa+o0ygAdEFwSKiOzv7miDxENBIAuqwAXHIECsJRMXwe3onYKV4w8QDcmJgEYAD2NMBQoJBQAOrVpDAQYAY0BsDVlLIA5AC0-VAAPlD9APoj4yP9pRhlGPXQAJLExHwAwgZcnDvkEDCijACCNF0gKAAKncCUaUKUNIaklcTwIFAGxCBe8f5ri8AG6ISoAV3ItxqUmkrGh93kT28qhsdigADFEDRqAjXP4ggEtDoAEoQLhOYhiBHkPJYLCeLE44AIpGKKDgsgQTT0CCkVj0zz9foCrApJm4u5s4TC0WCqDrTY0HZ7eAHI4nBjnS4oMkU+jUu7kQEgsGQhESOViqCooEQUEQqFSx7s4VQRnYyU1KApHZgZAGJAALwgKAlLLuEmsiu2u32BkOxzoZwuBiuesphpqxsxnojMNcApSwoWmBWUBV8cTmu1aZuRulz1e70+ImTDH+sPhDZdwi8qOJ-HDeKUBKCg6gGYNjBpdIZUAAqsQAMYGf3AQOIEMoVHDu7Rjax1XqpNiWvp8mZmdG-x7gtKS3Ya2bYF8UuLZbgaCVtUJ0MAaVhKB-0bNsxHdCs41-at23PFAQmQFAQN7SgFzASAaFXQ4kKkTwABlKgAdz4LCAKkFJ-38ZoaFadpOm6XoJHI4D33LH91VIIFKgw4BEAgcgOHaFAADVA3BCAgNE+BxNAgAhSpKngQxiEwKA1IgqTxOSKBNOgFDMU5ZdePeHJdO03TQLgicF0fec4PYv9OJobi+F4-jBLAFAbNslILP0tpUGs2zPAClAHMOJyXJqPiBIgITvJ9KA1AFLwQPoKAAGsIBASpNB0sToATSDjz-HD4l9KCOK4niYo8kSCtS+JbOMd9MHoWJNAMZdoAAOXeDEDCypLWCIahi1G4ARQwYwlnavhOu6qA+uIAasoAJmG7AJvG-jJtYYECoAeU0FIAAoAEoZCkAAjBSlJ+TAZo-OaaAW6BVogeTFOUzasAO6SIGOs7LukG67uUx7ZuIDquugABRfBdgQCB7C-Tb1KgW7FJSL77uIFhsCx+A1vFQbPvBh7CbugBmFJlo+1giYAFjp-qybWxm7oAVhSGAQC4W74Ehj82Kgrc+QRpGlJkYqqz5arXNquLPMlrhkdRyBHyAA).

A `Value extends Boolean` matches not only `Boolean`, but also the `NonFake2` and `FakeBoolean` in that example:

```typescript
interface NonFake {
	test: 'test'
}

interface NonFake2 {
	test: 'test'
	valueOf: () => boolean
}

interface FakeBoolean {
	valueOf: () => boolean
}
```

In other words: Any object that implements `valueOf: () => boolean`.

Because of that we can't really add these basic ones in my opinion, but we can add the more complex ones which are less likely to cause false positives.

**Comment by David245M** [comment]
> I think easiest solution here would be to add a new `exclude` parameter like `CamelCasedPropertiesDeep<ExampleType, String | Date>`

Did you mean create second optional type parameter that will be exceptional to deep copy?

**Comment by voxpelli** [comment]
> > I think easiest solution here would be to add a new `exclude` parameter like `CamelCasedPropertiesDeep<ExampleType, String | Date>`
> 
> Did you mean create second optional type parameter that will be exceptional to deep copy?

Yes, because then one adds it when one only adds it when one knows it won't create any false matches.

**Comment by David245M** [comment]
@voxpelli https://github.com/sindresorhus/type-fest/pull/327

**Comment by sindresorhus** [comment]
It should use the same solution as done here: https://github.com/sdotson/type-fest/blob/da1a717790396afb0fab16dfc61f6d3df5d0b6fb/source/delimiter-cased-properties-deep.d.ts#L48

**Comment by Emiyaaaaa** [comment]
Fixed on #890
